### PR TITLE
http_toolkit_v1

### PR DESCRIPTION
**Pull request by kretatusha**
The httptoolkit library is a tool for working with HTTP requests in Python that allows you to easily create, customize, and send requests to various services. It provides a simple interface for working with HTTP clients. The library also allows flexible transport customization and abstraction from a particular implementation of HTTP clients.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,168 @@
-# http_toolkit
\ No newline at end of file
+# http_toolkit
+The httptoolkit library is a tool for working with HTTP requests in Python that allows you to easily create, customize, and send requests to various services. It provides a simple interface for working with HTTP clients. The library also allows flexible transport customization and abstraction from a particular implementation of HTTP clients.
+
+## HTTPX_Service
```
Comment (SharUpOff) [inline README.md]: Do we need an underscore here?

File: README.md [hunk 2]
```diff
@@ -1 +1,168 @@
...
+    headers=(Header(name="ServiceHeader", value="service-header", is_sensitive=False),),
+    transport=AsyncHttpxTransport(base_url="https://example.com:4321", proxies={"http://": "http://10.10.1.10:3128"}),
+    ## base_url in this case is passed to transport
+)
+```
+
+### Sending a request
+
+```python
+### Async
```
Comment (SharUpOff) [inline README.md]: It seems there are no async examples below

File: README.md [hunk 3]
```diff
@@ -1 +1,168 @@
...
+    headers=(Header(name="SuperSecret", value="big_secret", is_sensitive=True, create_mask=lambda value: value[-4:])),
+    params={"over": "the rainbow"},
+    body="Something",
+)
+
+# By request
+service.request(Request(method=HttpMethod.POST, body="Request", params={}, path=""))
+```
+
+### Sending special types
```
Comment (SharUpOff) [inline README.md]: It is not clear what is the meaning of "special" here

File: tests/httptoolkit/service/test_async_service.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+
+class DummyAsyncService(AsyncService):
+    pass
+
+
+@pytest.fixture()
+def async_service() -> AsyncService:
+    return DummyAsyncService(
+        transport=AsyncHttpxTransport(base_url="https://kontur.ru", proxies={"http://": "http://10.10.1.10:3128"}),
```
Comment (SharUpOff) [inline tests/httptoolkit/service/test_async_service.py]: It's better to replace real domain examples with "example.com" or similar.

File: tests/httptoolkit/service/test_async_service.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    method: str,
+    async_service: AsyncService,
+    httpx_mock: HTTPXMock,
+):
+    response_text = "Schwarzenegger is a woman!"
+    path = "/some/path"
+
+    httpx_mock.add_response(
+        method=method.upper(),
+        url="https://kontur.ru" + path,
```
Comment (SharUpOff) [inline tests/httptoolkit/service/test_async_service.py]: It's better to replace real domain examples with "example.com" or similar.

File: tests/httptoolkit/service/test_async_service.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("method", ("get", "post", "put", "delete", "patch"))
+async def test_requests_when_successful(
+    method: str,
+    async_service: AsyncService,
+    httpx_mock: HTTPXMock,
+):
+    response_text = "Schwarzenegger is a woman!"
```
Comment (SharUpOff) [inline tests/httptoolkit/service/test_async_service.py:50]: Are we sure that no one will be offended by this joke? :')

File: tests/httptoolkit/service/test_async_service.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+    path = "/some/path"
+
+    httpx_mock.add_response(
+        method=method.upper(),
+        url="https://kontur.ru" + path,
```
Comment (SharUpOff) [inline tests/httptoolkit/service/test_async_service.py]: It's better to replace real domain examples with "example.com" or similar.

File: tests/httptoolkit/service/test_async_service.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+
+
+@pytest.mark.asyncio
+async def test_post_when_request_successful(async_service: AsyncService, httpx_mock: HTTPXMock):
+    response_text = "Schwarzenegger is a woman!"
+    path = "/some/path"
+
+    httpx_mock.add_response(
+        method="POST",
+        url="https://kontur.ru" + path + "?foo=bar",
```
Comment (SharUpOff) [inline tests/httptoolkit/service/test_async_service.py]: It's better to replace real domain examples with "example.com" or similar.

File: tests/httptoolkit/service/test_async_service.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
+
+@pytest.mark.asyncio
+async def test_if_response_not_successful(async_service: AsyncService, httpx_mock: HTTPXMock):
+    response_text = "Authorize first"
+    path = "/some/path"
+
+    httpx_mock.add_response(
+        method="GET",
+        url="https://kontur.ru" + path + "?foo=bar",
```
Comment (SharUpOff) [inline tests/httptoolkit/service/test_async_service.py]: It's better to replace real domain examples with "example.com" or similar.

File: tests/httptoolkit/service/test_async_service.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+    assert calls[0].headers["RequestHeader"] == "request-header"
+
+
+@pytest.mark.asyncio
+async def test_post_when_transport_error(async_service: AsyncService, httpx_mock: HTTPXMock):
+    path = "/some/path"
+
+    httpx_mock.add_exception(
+        method="POST",
+        url="https://kontur.ru" + path + "?foo=bar",
```
Comment (SharUpOff) [inline tests/httptoolkit/service/test_async_service.py]: It's better to replace real domain examples with "example.com" or similar.

File: tests/httptoolkit/service/test_async_service.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+@pytest.mark.asyncio
+@pytest.mark.parametrize("method", ("post", "put", "delete", "patch"))
+async def test_json_request_with_falsy_values(
+    async_service: AsyncService,
+    httpx_mock: HTTPXMock,
+    method: str,
+) -> None:
+    httpx_mock.add_response(
+        method=method.upper(),
+        url="https://kontur.ru/",
```
Comment (SharUpOff) [inline tests/httptoolkit/service/test_async_service.py]: It's better to replace real domain examples with "example.com" or similar.

File: tests/httptoolkit/service/test_async_service.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+@pytest.mark.asyncio
+@pytest.mark.parametrize("method", ("post", "put", "delete", "patch"))
+async def test_json_request_with_none_value(
+    async_service: AsyncService,
+    httpx_mock: HTTPXMock,
+    method: str,
+) -> None:
+    httpx_mock.add_response(
+        method=method.upper(),
+        url="https://kontur.ru/",
```
Comment (SharUpOff) [inline tests/httptoolkit/service/test_async_service.py]: It's better to replace real domain examples with "example.com" or similar.

File: README.md (hunk 1 above)
Comment (kretatusha) [reply README.md]: I guess not

File: README.md (hunk 2 above)
Comment (kretatusha) [reply README.md]: Yeah, I think I will remove this line

File: README.md (hunk 3 above)
Comment (kretatusha) [reply README.md]: change topic to `Sending JSON and multipart-files`

File: tests/httptoolkit/service/test_async_service.py (hunk 4 above)
Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py]: fixed

File: tests/httptoolkit/service/test_async_service.py (hunk 5 above)
Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py]: fixed

File: tests/httptoolkit/service/test_async_service.py (hunk 8 above)
Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py]: fixed

File: tests/httptoolkit/service/test_async_service.py (hunk 9 above)
Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py]: fixed

File: tests/httptoolkit/service/test_async_service.py (hunk 10 above)
Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py]: fixed

File: tests/httptoolkit/service/test_async_service.py (hunk 11 above)
Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py]: fixed

File: tests/httptoolkit/service/test_async_service.py (hunk 12 above)
Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py]: fixed

File: tests/httptoolkit/service/test_async_service.py (hunk 6 above)
Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py:50]: Do you think so? 😅

Comment (SharUpOff) [reply tests/httptoolkit/service/test_async_service.py:50]: Sure :')

Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py:50]: removed it 😢

Comment (kretatusha) [reply tests/httptoolkit/service/test_async_service.py:50]: so i returned it back

Comment (SharUpOff) [reply tests/httptoolkit/service/test_async_service.py:50]: With the hope for a healthy humor :'D

**Review by SharUpOff — APPROVED** [review]
Have no more questions :)

**Pull request merged**